_compdemos/Cromwell.md

```diff
@@ -4,18 +4,101 @@ primary_reviewers: vortexing
+
+While workflows are being developed in the context of a WDL for running with Cromwell (or other WDL runners that are being developed), each task must be defined to occur in a specific environment.  That could be a user's environment on the local SLURM cluster, or that could be a clean environment with specific software packages loaded, or it could be a docker or singularity container that directly specifies the software and environment in which to run the task. 
+
+#### EasyBuild and Lmod
+One approach to making software versioning and environment management easier and available to users on a shared HPC cluster like our SLURM cluster, is the use of a "module" system.  We have more information in this Wiki on [managing your software environment at the Hutch](/scicomputing/compute_environments/), but to summarize here, essentially two users can load a single, defined module when they want to run a task and the versions of all the software required for that module to work will be loaded identically for either user.  That means that regardless of who is running a task, all users can load the same version of a software package to ensure that they get the same results!  For those of us here at the Hutch who have access to a [wide variety of software modules](/scicomputing/compute_scientificSoftware/) (built via [EasyBuild](https://github.com/FredHutch/easybuild-life-sciences) for us!), this is a useful and time-saving system that can very easily be leveraged using our configuration of Cromwell. 
+
+#### Docker and Singularity
+For all of those, however, who do not have access to our SLURM cluster and all those magic software modules, it means that for them, our workflows still aren't directly portable or shareable.  For that scenario, a containerization approach like using [Docker](/scicomputing/compute_environments/#docker-containers).  What this looks like in the context of workflows is that every task described in a workflow is given as a runtime variable a Docker container name and tag which specifies the exact container to run the task inside that will have all the software required for the task to be performed.  In the case of workflows running on our local SLURM cluster, that means that Cromwell is configured to use Singularity in order to pull and convert Docker images prior to running the task.  Meanwhile, on cloud based backends like AWS Batch, it means that the Docker container can be pulled and used without conversion.  Designing workflows to use publicly available Docker containers provides a substantial improvement in the likelihood that another user or collaborator would be able to run your workflow, reproducibly, on our or their own HPC resources.  
```

Comment:
For all of those, however, who do not have access to our SLURM cluster and all those magic software modules, our workflows are not directly portable or shareable.  For that scenario, a containerization approach like using [Docker](/scicomputing/compute_environments/#docker-containers) would be more appropriate.  What this looks like in the context of workflows is that every task described in a workflow is given as a runtime variable, a Docker container name and tag, which specifies the exact container to run the task inside that will have all the software required for the task to be performed.  In the case of workflows running on our local SLURM cluster, Cromwell is configured to use Singularity in order to pull and convert Docker images prior to running the task.  Meanwhile, on cloud based backends like AWS Batch, it means that the Docker containers can be pulled and used without conversion.  Designing workflows to use publicly available Docker containers provides a substantial improvement in the likelihood that another user or collaborator would be able to run your workflow, reproducibly, on our or their own HPC resources.